Add plot functions from audsp

frankenjoe opened the pull request:
Adds the following plot functions from `audsp`:

* cepstral()

  ![image](https://user-images.githubusercontent.com/10383417/125396401-c8f03e80-e3ac-11eb-8dc2-d39495204617.png)

* signal()

  ![image](https://user-images.githubusercontent.com/10383417/125300576-766e3e00-e32a-11eb-97ee-a8aaa443d710.png)

* spectrum()
  
  ![image](https://user-images.githubusercontent.com/10383417/125396422-d1e11000-e3ac-11eb-981c-d2eb191ce0ff.png)

frankenjoe commented:
@hagenw I'm using `librosa` in the docstring examples, but `pytest` is complaining that `librosa` is not defined even though I added it to the requirements of `tests` and `docs` and do e.g.:

```
    Example:
        .. plot::
            :context: reset
            :include-source: false

            import librosa
            from audplot import signal

        .. plot::
            :context: close-figs

            >>> x, sr = librosa.load(librosa.ex('trumpet'))
            >>> dur = len(x) / sr
            >>> signal(x, dur)
```

Any idea how to fix this?

frankenjoe commented:
> Any idea how to fix this?

I now changed the examples to:

```
    Example:
        .. plot::
            :context: reset
            :include-source: false

            from audplot import signal

        .. plot::
            :context: close-figs

            >>> import librosa
            >>> x, sr = librosa.load(librosa.ex('trumpet'))
            >>> dur = len(x) / sr
            >>> signal(x, dur)
```

which seems to work.

frankenjoe commented:
Tests are passing now, this is ready for review.

hagenw commented on docs/api.rst:
```diff
@@ -22,3 +27,13 @@ series
+signal
+------
+
+.. autoclass:: signal
```
This should be `autofunction`. Interestingly, it does not fail this way, but it produces

![image](https://user-images.githubusercontent.com/173624/125251899-16f73a80-e2f8-11eb-8ef8-84930d244ae9.png)

frankenjoe replied:
I copied and paste from `series`, so I will change it for `series`, too.

frankenjoe replied:
done

hagenw commented on docs/api.rst:
```diff
@@ -22,3 +27,13 @@ series
+spectrum
+--------
+
+.. autoclass:: spectrum
```
The same as for `signal()`, please change to `autofunction`

frankenjoe replied:
done

hagenw commented on audplot/core/api.py:
```diff
@@ -190,3 +258,115 @@ def series(
+
+def signal(
+        x: np.ndarray,
+        dur: float,
```
I would change `dur` to `duration`

frankenjoe replied:
done

hagenw commented on audplot/core/api.py:
```diff
@@ -190,3 +258,115 @@ def series(
+
+
+def spectrum(
+        mag: np.ndarray,
```
I would change `mag` to `magnitude`

hagenw commented on audplot/core/api.py:
```diff
@@ -190,3 +258,115 @@ def series(
+
+def spectrum(
+        mag: np.ndarray,
+        dur: float,
```
I would change `dur` to `duration`

frankenjoe replied:
done

hagenw reviewed (commented):
Thanks for adding more plots.
You should maybe also update the README as it states that it mainly targets machine learnibng problems:

> audplot includes plotting functions targeted at machine learning problems like a confusion matrix.

I think, I added it as before we only had `truth` and `prediction` as inputs to plotting functions.

hagenw commented on audplot/core/api.py:
```diff
@@ -8,6 +8,67 @@
+        channel: int = 0,
+        num_ticks: int = 10,
+        ax: plt.Axes = None,
+        cmap: str = 'jet',
```
BTW, the y-axis of the spectrum plot is cut and should be fixed.

hagenw replied:
And I would vote for showing a colorbar with actual dB values next to the plot to easier judge what the plot is showing.

frankenjoe replied:
I changed to `magma ` and fixed the `y-axis`. Will take care about the colorbar in another commit.

frankenjoe replied:
`cepstrum()` and `spectrum()` now return the image object, which can be used to draw the colorbar, e.g.:

```python
...
image = cepstrum(y, dur)
plt.colorbar(image)
plt.show()
```

The problem of directly plotting it in the function is that the colorbar is plotted to a different axis. But I think the current solution should be fine.

hagenw replied:
Could you add `dB` after the numbers in the colormap.

And I wonder if the range of dB values for cepstrum makes sense

![image](https://user-images.githubusercontent.com/173624/125260644-d354fe80-e300-11eb-9a74-7ecf8d4f8d88.png)

frankenjoe replied:
> And I wonder if the range of dB values for cepstrum makes sense

Probably not :) Unfort in the documentation of `librosa` they don't label the colorbar: https://librosa.org/doc/main/generated/librosa.feature.mfcc.html

frankenjoe replied:
But I now added `db` to the `spectrum()` example (see plot in description).

hagenw replied:
OK, thanks, then let's stay with the current behavior for `cepstrum()`

frankenjoe commented on audplot/core/api.py:
```diff
@@ -8,6 +8,67 @@
 
 
+def cepstrum(
+        cc_matrix: np.ndarray,
```
cepstral coefficients

hagenw replied:
OK, maybe integrate that information into the docstring:

```rst
Args:
    cc_matrix: cepstral coefficients matrix with magnitude values
```

frankenjoe replied:
done

hagenw commented on audplot/core/api.py:
```diff
@@ -8,6 +8,67 @@
+    ts_sec = ["{:4.2f}".format(i) for i in np.linspace(0, dur, num_ticks)]
+    ax.set_xticks(ts)
+    ax.set_xticklabels(ts_sec)
+    ax.set_xlabel("Time (sec)")
```
I personally prefer `Time / s` or `Time / sec`, but if you think using `()` is more consistent with what others do in audEERING it's also fine with me.

frankenjoe replied:
Don't have a strong opinion here, so I changed to `Time / s` and `Frequency / Hz`

frankenjoe commented:
> You should maybe also update the README

![image](https://user-images.githubusercontent.com/10383417/125260527-b28ca900-e300-11eb-9af6-7b524e0369af.png)

hagenw commented:
The duration of the cepstrum and spectrum plot has some strange xtick values:

![image](https://user-images.githubusercontent.com/173624/125282736-0060db80-e318-11eb-96d4-4a7dcd7bf6b2.png)

frankenjoe commented:
> The duration of the cepstrum and spectrum plot has some strange xtick values:

why do you think it's strange?

hagenw commented:
> > The duration of the cepstrum and spectrum plot has some strange xtick values:
> 
> why do you think it's strange?

I would expect something like:
```
0, 0.5, 1.0, ...
```
not
```
0, 0.59, 1.19, ...
```

frankenjoe commented:
The user can control the number of ticks with the `num_ticks` argument, but the actual numbers depend of course on the duration of the file. if the duration is an odd number, the ticks get also odd numbers. Maybe we should also let the user control the number of decimal points and set to 1 by default?

hagenw commented:
> The user can control the number of ticks with the `num_ticks` argument, but the actual values depend of course on the duration of the file. if the duration is an odd, the ticks get also odd values. Maybe we should also let the user control the number of decimal points and set to 1 by default?

I would remove the `num_ticks` and go with the one that `matplotlib` decides to print. We only need to convert them to duration. For that I would maybe replace the `duration` argument by a `sampling_rate` argument as originally I thought with `duration` you can limit the presented signal to the given duration.

frankenjoe commented:
> For that I would maybe replace the `duration` argument by a `sampling_rate` argument as originally

I think we had a strong reason to have a `duration` and not a `sampling_rate` argument. Maybe it was because of `spectrum()` and `cepstrum()` functions, where you have to calculate the sampling rate from the hop duration, but I'm not completely sure any more. So, I'll give `sampling_rate` another chance, cause I also think it makes more sense. The reason for `num_ticks` was that `matplotlib` not always chooses them very smartly I think. But I'm also fine with removing them.

hagenw commented:
> The reason for `num_ticks` was that `matplotlib` not always chooses them very smartly I think. But I'm also fine with removing the argument.

Then the user could still set the ticks herself. But my hope is that results will be better as they are at the moment.

frankenjoe commented:
> Then the user could still set the ticks herself. But my hope is that results will be better as they are at the moment.

Agree.

I replaced the `duration` argument with `sampling_rate` and removed the `num_ticks` argument. I updated the plots in the descriptions.

hagenw commented on audplot/core/api.py:
```diff
@@ -8,6 +9,73 @@
+            >>> import matplotlib.pyplot as plt
+            >>> x, sr = librosa.load(librosa.ex('trumpet'))
+            >>> y = librosa.feature.mfcc(x, sr)
+            >>> y_sr = sr / 512  # default hop length is 512
```
OK, now I see what you mean. I don't think a user will ever realize that she has to do this.
So the advantage of giving duration was that you could calculate this without knowing the hop duration.

Maybe it will be more obvious if we add `hop_duration` as another argument?

Or at least we have to update the description of sampling rate, stating that it is not the sampling rate of the original signal.

frankenjoe replied:
Yes, and the problem with `duration` is that then it's not obvious how to rename the ticks (since we don't have the `sampling_rate`) and that was the reason to not use the default ticks but customize via `num_ticks`.

But maybe `hop_duration` is indeed a good compromise.

frankenjoe replied:
Switched to `hop_duration` now, hopefully the last time I had to update the plots in the description :)

hagenw replied:
Cool, looks good now.

hagenw commented:
Could you resolve the conflicts, then I will merge.

frankenjoe commented:
> Could you resolve the conflicts, then I will merge.

done
